engine/linksystem.go

```diff
@@ -83,30 +83,33 @@ func (e *Engine) mkLinkSystem() ipld.LinkSystem {
 			// If the link is not found, it means that the root link of the list has
 			// not been generated and we need to get the relationship between the cid
 			// received and the contextID so the lister knows how to
-			// regenerate the list of CIDs.
-			key, err := e.getCidKeyMap(ctx, c)
+			// regenerate the list of CIDs. There can be multiple provider + context IDs associated with a cid
+			// as index provider supports publishing on behalf of multiple providers.
+			keys, err := e.getAllCidKeyMap(ctx, c)
```

Comment:
`e.getAllCidKeyMap` would essentially iterate over the entire ad chain right? 
I am worried this would be expensive to do per read.

Would it make sense to:
- benchmark this, or
- think of maintaining some alternative reverse mapping that offers a cheaper/faster lookup?